python/lsst/ip/diffim/imageDecorrelation.py

```diff
@@ -223,21 +233,32 @@ def run(self, exposure, templateExposure, subtractedExposure, psfMatchingKernel,
         return pipeBase.Struct(correctedExposure=correctedExposure, correctionKernel=corrKern)
 
     @staticmethod
-    def _computeDecorrelationKernel(kappa, svar=0.04, tvar=0.04):
+    def _computeDecorrelationKernel(kappa, svar=0.04, tvar=0.04, preConvKernel=None):
         """! Compute the Lupton/ZOGY post-conv. kernel for decorrelating an
         image difference, based on the PSF-matching kernel.
         @param kappa  A matching kernel 2-d numpy.array derived from Alard & Lupton PSF matching
         @param svar   Average variance of science image used for PSF matching
         @param tvar   Average variance of template image used for PSF matching
+        @param preConvKernel If not None, then pre-filtering was applied
+            to science exposure, and this is the pre-convolution kernel.
         @return a 2-d numpy.array containing the correction kernel
 
         @note As currently implemented, kappa is a static (single, non-spatially-varying) kernel.
         """
         kappa = DecorrelateALKernelTask._fixOddKernel(kappa)
-        kft = scipy.fftpack.fft2(kappa)
-        kft = np.sqrt((svar + tvar) / (svar + tvar * kft**2))
-        pck = scipy.fftpack.ifft2(kft)
-        pck = scipy.fftpack.ifftshift(pck.real)
+        kft = np.fft.fft2(kappa)
+        kft2 = np.conj(kft) * kft
+        denom = svar + tvar * kft2
+        if preConvKernel is not None:
+            mk = DecorrelateALKernelTask._fixOddKernel(kft)
+            mk[mk < 1e-3] = 1e-3  # Psf should not be < 0; and messes up denominator
```

Comment:
Please hard-coding the limit as a constant (e.g. `double MIN_KERNEL = 1.0e-3`) and document it where you define it. That keeps the code consistent in case you decide to change the limit -- the number only appears once.

Is 0.001 sufficiently small to be negligible? Some kernels can easily have 1000-ish pixels.

Given the code, I'm guessing kernel <= 0 messes up the denominator? Because if only negative values were a problem I would have expected the test to replace negative values with 0. If my supposition is correct, please update the comment, e.g. `...should not be <= 0...`.

Reply:
All good points. Yes, it seems that 1e-3 is small enough -- smaller and strangeness can happen in the FFTs.